Dockerfile

```diff
@@ -1,16 +1,74 @@
-FROM gcr.io/distroless/static:latest
-LABEL Name="s3gw-cosi-driver"
-LABEL maintainers="s3gw maintainers"
-LABEL description="s3gw COSI driver"
+#########################################################################################
+# Build
+#########################################################################################
 
+# First stage: building the driver executable.
+FROM docker.io/library/golang:1.22 as builder
+
+# Set the working directory.
+WORKDIR /work
+
+# Prepare dir so it can be copied over to runtime layer.
+RUN mkdir -p /var/lib/cosi
+
+# Copy the Go Modules manifests.
+COPY go.mod go.mod
+COPY go.sum go.sum
+
+# Cache dep before building and copying source so that we don't need to re-download as
+# much and so that source changes don't invalidate our downloaded layer.
+RUN go mod download
+
+# Copy the go source.
+COPY Makefile Makefile
+COPY cmd/ cmd/
+COPY pkg/ pkg/
+
+# Build.
+RUN make build
+
+#########################################################################################
+# Runtime
+#########################################################################################
+
+# Second stage: building final environment for running the executable.
+FROM gcr.io/distroless/static:latest AS runtime
+
+# Copy the executable.
+COPY --from=builder --chown=65532:65532 /work/bin/s3gw-cosi-driver /usr/bin/s3gw-cosi-driver
```

Comment:
can we get an explanation on where `65532` comes from, for posterity?

Reply:
This is just a convention taken from [GCR Distroless images](https://github.com/GoogleContainerTools/distroless/blob/main/common/variables.bzl) denoting a `nonroot` user.

Reply:
TIL :)